compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,622 @@
+    ArrayRef<int64_t> srcShape = extractOp.getSourceType().getShape();
+
+    // Permute the offsets, sizes, and strides to pre-transpose ordering.
+    applyPermutationToVector(offsets, invPerm);
+    applyPermutationToVector(sizes, invPerm);
+    applyPermutationToVector(strides, invPerm);
+    SmallVector<int64_t> baseShape = applyPermutation(srcShape, invPerm);
+
+    // Check if the resulting offsets, sizes, and strides correspond to a
+    // contiguous slice and can thus be mappable to a `flow.tensor.update` op.
+    // This should always be worth doing because this can remove a dispatch for
+    // the slice, and the transpose is on the slice rather than the full tensor.
+    if (!IREE::Flow::isOffsetSizeAndStrideMappableToFlow(offsets, sizes,
+                                                         strides, baseShape)) {
+      return failure();
+    }
+
+    ArrayRef<int64_t> staticSizes = extractOp.getStaticSizes();
+    ArrayRef<int64_t> sliceShape = extractOp.getResultType().getShape();
+    llvm::SmallDenseSet<unsigned> rankReducingMask =
+        *mlir::computeRankReductionMask(staticSizes, sliceShape);
+
+    int64_t dim = 0;
+    llvm::SmallDenseMap<int64_t, int64_t> rankReducedMap;
+    for (int64_t i = 0, e = perm.size(); i < e; ++i) {
+      if (!rankReducingMask.contains(i)) {
+        rankReducedMap[i] = dim++;
+      }
+    }
+
+    // Compute the new permutation by dropping all rank-reduced dimensions.
+    SmallVector<int64_t> rankReducedPerm;
+    for (int64_t i : perm) {
+      if (!rankReducingMask.contains(i)) {
+        rankReducedPerm.push_back(rankReducedMap[i]);
+      }
+    }
+
+    auto rankReducedInvPerm = invertPermutationVector(rankReducedPerm);
+
+    RankedTensorType sliceType = getPermutedTensorType(
+        cast<RankedTensorType>(extractOp.getType()), rankReducedInvPerm);
+    Value slice = rewriter.create<tensor::ExtractSliceOp>(
+        extractOp.getLoc(), sliceType, transposeOp.getInput(), offsets, sizes,
+        strides);
+    // Transpose back to the original slice.
+    if (!isIdentityPermutation(rankReducedPerm)) {
+      slice = createTranspose(rewriter, slice, rankReducedPerm);
+    }
+    rewriter.replaceOp(extractOp, slice);
+    return success();
+  }
+};
+
+// Sinks a transpose through a tensor.expand_shape.
+class SinkTransposeThroughExpandShape
+    : public OpRewritePattern<tensor::ExpandShapeOp> {
+public:
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expandOp,
+                                PatternRewriter &rewriter) const override {
+    Value source = expandOp.getSrc();
+    auto transposeOp = source.getDefiningOp<linalg::TransposeOp>();
+    // Do not propagate through reshapes if the transpose has multiple users, as
+    // this could end up duplicating the transposes. We should only propagate
+    // through reshape when it is free to do so.
+    if (!transposeOp || !transposeOp->hasOneUse()) {
```

Comment:
Wondering why this check isnt done for the slice case above?

Reply:
The slice case is a little less clear. I restricted it such that propagation only occurs if the resulting slice can be lowered to a `flow.update` op, meaning we save a dispatch by doing the propagation and still have an opportunity to fuse the transpose with a consumer. I'm figuring this is always worth doing, even if it means we end up with extra transposes. I can add the same check here to take it slower though.